File: Command/java/CommandPattern.java
package Command.java;

public class CommandPattern {

    public static void main(String[] args){
        Button button = new Button();

        Light light = new Light();

        button.setCommand(new TurnOffCommand(light));
        button.pressed();
        button.undoPressed();

        button.setCommand(new TurnOnCommand(light));
        button.pressed();
        button.undoPressed();
    }
}



// Invoker
class Button {
    private Command command;

    public void setCommand(Command command) {
        this.command = command;
    }

    public void pressed() {
        command.execute();
    }

    public void undoPressed() { command.undo(); }
}

// Receiver
class Light {

    public void turnOn() {
        System.out.println("***** Light is ON *****");
    }

    public void turnOff() {
        System.out.println("_____ Light is OFF _____");
    }

    public void undo() {
        System.out.println("%%%%% UNDO %%%%%");
    }
}

// Command
interface Command {
    void execute();
    void undo();
}

class TurnOnCommand implements Command {

    private Light light;

    public TurnOnCommand(Light light) {
        this.light = light;
    }

    @Override
    public void execute() {
        light.turnOn();
    }

    @Override
    public void undo() {
        light.undo();
    }
}

class TurnOffCommand implements Command {

    private Light light;

    public TurnOffCommand(Light light) {
        this.light = light;
    }

    @Override
    public void execute() {
        light.turnOff();
    }

    @Override
    public void undo() {
        light.undo();
    }
}

File: Command/javascript/CommandPattern.js
// Invoker
var Button = function(){
    var _command;

    this.setCommand = function(command){
        if (command instanceof Command)
            _command = command;
        else
            throw new Error("command should be instanceof Command");
    };

    this.pressed = function(){
        _command.execute();
    };
};



// Receiver
var Light = function(){
    this.turnOn = function(){
        console.log("***** Light is ON *****");
    };
    this.turnOff = function(){
        console.log("_____ Light is OFF _____");
    };
};



// Command
var Command = function(light){
    if (light instanceof Light)
        this.light = light;
    else
        throw new Error("light should be instanceof Light");
};
Command.prototype.execute = function(){
    // JavaScript에서는 부모의 메서드를 override하도록
    // 강제할 수 있는 방법이 없다.
    // 따라서 override 안된 경우 강제로 에러 처리
    throw Error("This method should be overridden");
};

var TurnOnCommand = function(light){
    Command.call(this, light);
};
TurnOnCommand.prototype = Object.create(Command.prototype);
TurnOnCommand.prototype.constructor = TurnOnCommand;
TurnOnCommand.prototype.execute = function(){
    this.light.turnOn();
};

var TurnOffCommand = function(light){
    Command.call(this, light);
};
TurnOffCommand.prototype = Object.create(Command.prototype);
TurnOffCommand.prototype.constructor = TurnOffCommand;
TurnOffCommand.prototype.execute = function(){
    this.light.turnOff();
};



(function(){

    var button = new Button();

    var light = new Light();

    button.setCommand(new TurnOffCommand(light));
    button.pressed();

    button.setCommand(new TurnOnCommand(light));
    button.pressed();

})();
